name: Testing

on:
  push:
    branches:
      - '**'
  pull_request:

jobs:
  test:
    name: Ruby ${{ matrix.ruby }} (Rack ${{ matrix.rack }})
    runs-on: ubuntu-latest
    timeout-minutes: 15
    strategy:
      fail-fast: false
      matrix:
        rack:
          - stable
        # Due to https://github.com/actions/runner/issues/849, we have to use quotes for '3.0'
        ruby: [2.6, 2.7, '3.0', 3.1, 'jruby-9.3', truffleruby, truffleruby-head]
        include:
          - ruby: 3.1
            env:
              BUNDLE_GEMFILE: gems/rack-v3.rb
            allow-failure: true
          - ruby: jruby-head
            env:
              BUNDLE_GEMFILE: gems/rack-v2.rb
            allow-failure: true
    steps:
    - uses: actions/checkout@v3
    - uses: ruby/setup-ruby@v1
      continue-on-error: ${{ matrix.allow-failure || false }}
      id: setup-ruby
      with:
        ruby-version: ${{ matrix.ruby }}
        bundler-cache: true
    - name: Install dependencies
      run: sudo apt-get install -y pandoc nodejs pkg-config libxml2-dev libxslt-dev
    - name: Run sinatra tests
      continue-on-error: ${{ matrix.allow-failure || false }}
      id: tests
      run: bundle exec rake
    - name: Run sinatra-contrib tests
      continue-on-error: ${{ matrix.allow-failure || false }}
      id: contrib-tests
      working-directory: sinatra-contrib
      run: |
        bundle install --jobs=3 --retry=3
        bundle exec rake
    - name: Run rack-protection tests
      continue-on-error: ${{ matrix.allow-failure || false }}
      id: protection-tests
      working-directory: rack-protection
      run: |
        bundle install --jobs=3 --retry=3
        bundle exec rake
    - uses: 8398a7/action-slack@v3
      with:
        status: ${{ job.status }}
        fields: repo,message,commit,author,ref,job,took # selectable (default: repo,message)
      env:
        SLACK_WEBHOOK_URL: ${{ secrets.SLACK_WEBHOOK }} # required
        MATRIX_CONTEXT: ${{ toJson(matrix) }} # required
      if: failure() && env.SLACK_WEBHOOK_URL
    # because continue-on-error marks the steps as pass even if they fail
    - name: "setup-ruby (bundle install) outcome: ${{ steps.setup-ruby.outcome }}"
      run: ""
    - name: "sinatra tests outcome: ${{ steps.tests.outcome }}"
      run: ""
    - name: "sinatra-contrib tests outcome: ${{ steps.contrib-tests.outcome }}"
      run: ""
    - name: "rack-protection tests outcome: ${{ steps.protection-tests.outcome }}"
      run: ""
